### DISCUSSION TITLE
C++ Client Libraries for Dialogflow

### DISCUSSION THREAD

**Discussion by DrFrederick** [googleapis/google-cloud-cpp#9499] [General]
I am intended to write a code for voice analytic using Google Cloud CPP Dialogflow. Hence I am writing a piece of code to test the CreateConversation() and  CreateParticipant().

When testing I found CreateConversation() is working but the CreateParticipant() is throwing error. 

At this juncture I am not able to identify where is my mistake. I am keen on looking for someone's help to solve this issues.

Here is my source code
[ClientConversation.txt](https://github.com/googleapis/google-cloud-cpp/files/9129358/ClientConversation.txt)

Here is the output
admin@ip-172-31-7-209:~/gcloudagent/build$ ./gcloudagent
 Main Start!
Google Credentials : /home/admin/pratheepagentassist-e385115eac3d.json
 Profile : projects/pratheepagentassist/conversationProfiles/D7ILDLMdSe63ud_nOkFtUA
 Stage : 2
 Name : projects/pratheepagentassist/conversations/119gKpF0rQcSkSE3ijs79JLfw
 Parent : projects/pratheepagentassist/conversations/119gKpF0rQcSkSE3ijs79JLfw

 Before CreateParticipant() !
E0718 03:59:11.057965025   21520 client_context.cc:115]      assertion failed: call_ == nullptr
Aborted
admin@ip-172-31-7-209:~/gcloudagent/build$

**Comment by coryan** [comment]
You are reusing a `grpc::ClientContext` across two RPCs, this is no allowed by the gRPC library:

> "[ClientContext](https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_context.html) instances should not be reused across rpcs.

https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_context.html#details

Consider using the C++ client libraries (you are using what we would call the generated gRPC stubs).  Good starting points may be:

https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/dialogflow_cx/quickstart/README.md
https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/dialogflow_es/quickstart/README.md

**Reply by coryan** [reply]
Before you go too deep into using `v2beta1`, be aware that these version of the API is not GA.  As far as I know, there are no plans to make that API GA either.  Please consider moving to the v3 API.

**Reply by DrFrederick** [reply]
> You are reusing a `grpc::ClientContext` across two RPCs, this is no allowed by the gRPC library:
> 
> > "[ClientContext](https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_context.html) instances should not be reused across rpcs.
> 
> https://grpc.github.io/grpc/cpp/classgrpc_1_1_client_context.html#details
> 
> Consider using the C++ client libraries (you are using what we would call the generated gRPC stubs). Good starting points may be:
> 
> https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/dialogflow_cx/quickstart/README.md https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/dialogflow_es/quickstart/README.md

Hi Coryan,
Thanks for your valuable suggestion, 
As per your suggestion I have changed the source with single context for the RPC but program aborted, can you please review again.

 I have attached the new source  
[ClientConversation-2.txt](https://github.com/googleapis/google-cloud-cpp/files/9142127/ClientConversation-2.txt)

Output is 
admin@ip-172-31-7-209:~/gcloudagent/build$ ./gcloudagent                         Main Start!
Google Credentials : /home/admin/pratheepagentassist-e385115eac3d.json
 Profile : projects/pratheepagentassist/conversationProfiles/D7ILDLMdSe63ud_nOkFtUA
 Stage : 2
 Name : projects/pratheepagentassist/conversations/081be8OtlSgQfCvrLqxvgWNhQ
 Parent : projects/pratheepagentassist/conversations/081be8OtlSgQfCvrLqxvgWNhQ

 Before CreateParticipant() !
E0719 14:34:23.372970638   25886 call_op_set.h:980]          assertion failed: false
Aborted
admin@ip-172-31-7-209:~/gcloudagent/build$

**Reply by DrFrederick** [reply]
> Before you go too deep into using `v2beta1`, be aware that these version of the API is not GA. As far as I know, there are no plans to make that API GA either. Please consider moving to the v3 API.

Yes I can use v3 API, Since I am looking to implement the voice analysis application,  especially I need the StreamingAnalyze RPC interfaces.
human_agent_assistant.proto
ivr_node.proto
operations.proto
phone_number.proto
phone_number_order.proto

Can you please let me know when can I get the v3 API.

**Reply by coryan** [reply]
The `v3` API is available right now.  I linked the quickstart for the C++ client library:

https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/dialogflow_cx/quickstart/README.md

The raw protos are available from:

https://github.com/googleapis/googleapis/tree/master/google/cloud/dialogflow/cx/v3

I think you may be looking for this:

https://github.com/googleapis/googleapis/blob/53377c165584e84c410a0905d9effb3fe5df2806/google/cloud/dialogflow/cx/v3/session.proto#L82-L83

Which corresponds to:

https://github.com/googleapis/google-cloud-cpp/blob/ba7158d676c076b074211f9cfc35b17a3c9501f5/google/cloud/dialogflow_cx/sessions_client.h#L116-L119

Keep in mind that I am not an expert in the dialogflow APIs.

**Reply by coryan** [reply]
Regarding the crash.  Can you share what version of gRPC are you using?  The assertion is probably from somewhere around here:

https://github.com/grpc/grpc/blob/4ead675492d87c155a4b1d32a3000471f0468167/include/grpcpp/impl/codegen/call_op_set.h#L979-L986

Having the exact version would help.

**Reply by coryan** [reply]
I think you may be asking about this feature:

https://github.com/googleapis/googleapis/commit/b9927eb6fe36783981201eecdf0b95aa340c9137

as best as I can tell that feature is not GA, and not available in the `v2` or `v3` versions of the API.  I do not have a good answer for you at this time.  Give me a few days to get back to you.

**Reply by coryan** [reply]
FWIW, the next version of `google-cloud-cpp` will include `StreamingAnalyzeContent` as part of the v2 API.

https://github.com/googleapis/google-cloud-cpp/blob/2ae1f558bd18dde0946c0238f2d0b2301bee4f71/google/cloud/dialogflow_es/participants_client.h#L358-L361
